agroal-pool/src/main/java/io/agroal/pool/util/StampedCopyOnWriteArrayList.java

```diff
@@ -119,33 +119,29 @@ public T removeLast() {
 
     @Override
     public boolean remove(Object element) {
-        long stamp = lock.readLock();
+        int index = indexOf( element );
+        if ( index == -1 ) {
+            return false;
+        }
+        long stamp = lock.writeLock();
         try {
-            boolean found = true;
-
-            while ( found ) {
-                T[] array = data;
-                found = false;
-
-                for ( int index = array.length - 1; index >= 0; index-- ) {
-                    if ( element == array[index] ) {
-                        found = true;
-
-                        long writeStamp = lock.tryConvertToWriteLock( stamp );
-                        if ( writeStamp != 0 ) {
-                            stamp = writeStamp;
-
-                            T[] newData = Arrays.copyOf( data, data.length - 1 );
-                            arraycopy( data, index + 1, newData, index, data.length - index - 1 );
-                            data = newData;
-                            return true;
-                        } else {
-                            break;
-                        }
+            if ( element != data[index] ) {
```

Comment:
That's exactly the point! If `data[index]` is not the element we want to remove anymore, need to find that element again. (by the way, can't call `indexOf` here because holding a write lock now).

Reply:
Yeah, but removing the 2nd to last element followed by the last element would lead to an exception due to the new size of the array

Reply:
ok, this needs to change to `if ( index >= data.length || element != data[index] )`

Reply:
Write lock at the start instead ? That would simplify the code a lot

Reply:
Most times the scan outside the loop will hold, greatly reducing the lock window. I think the extra complexity is worth in this case.

Reply:
JMH benchmark ?